# ISCS 30.18 Exercise 3 - Data Cleaning

`Group: JM Amador, Jet Bana, Gio Divino, Sid Vincent Paz, Franz Taborlupa`

##Steps taken for cleaning:
* Changed the data type of all columns containing dates from object to datetime64
* Fixed the inconsistent capitalization of the Gender column
* Dropped irrelevant columns Source and Source_II since these contain links irrelevant to the problem
* Fixed stray values for the Travel_history_location and Confirmation_method columns since some values have the same meaning but are written differently
* Fixed separation format for for Travel_history_country
* Fixed stray values in Travel_history_entry column and converted the data type to datetime64

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('data.csv')
pd.set_option('display.max_columns', None)

In [ ]:
df.head(3) # check the dataset and how it looks

ID     Status                             Location    City  Country  Age  \
0   1  confirmed  Guy's and St Thomas Hospital London  London  England  NaN   
1   2  confirmed  Guy's and St Thomas Hospital London  London  England  NaN   
2   3  confirmed                               London  London  England  NaN   

  Gender  Date_onset Date_confirmation        Symptoms Hospitalised (Y/N/NA)  \
0    NaN  2022-04-29        2022-05-06            rash                     Y   
1    NaN  2022-05-05        2022-05-12            rash                     Y   
2    NaN  2022-04-30        2022-05-13  vesicular rash                     N   

  Date_hospitalisation Isolated (Y/N/NA) Date_isolation Outcome Date_death  \
0           2022-05-04                 Y     2022-05-04     NaN        NaN   
1           2022-05-06                 Y     2022-05-09     NaN        NaN   
2                  NaN                 Y            NaN     NaN        NaN   

                   Contact_comment  Contact_ID Contact_location  \
0                              NaN         NaN              NaN   
1  Index Case of household cluster         3.0        Household   
2                              NaN         2.0        Household   

  Travel_history (Y/N/NA) Travel_history_entry Travel_history_start  \
0                       Y           2022-05-04           late April   
1                       N                  NaN                  NaN   
2                       N                  NaN                  NaN   

  Travel_history_location Travel_history_country   Genomics_Metadata  \
0  Lagos and Delta States                Nigeria  West African Clade   
1                     NaN                    NaN  West African Clade   
2                     NaN                    NaN  West African Clade   

  Confirmation_method                                             Source  \
0              RT-PCR  https://www.gov.uk/government/news/monkeypox-c...   
1              RT-PCR  https://www.gov.uk/government/news/monkeypox-c...   
2              RT-PCR  https://www.gov.uk/government/news/monkeypox-c...   

                                           Source_II  Date_entry  \
0  https://www.who.int/emergencies/disease-outbre...  2022-05-18   
1                                                NaN  2022-05-18   
2                                                NaN  2022-05-18   

  Date_last_modified  
0         2022-05-18  
1         2022-05-18  
2         2022-05-18

In [ ]:
df.info() 

# Seeing lots of null values but there is a certain level of uncertainty with this virus and data gathering due to the nature of the disease. 
# Dropping rows with null values eliminates highly valuable data seen elsewhere in the row therefore we decided to retain all null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65777 entries, 0 to 65776
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       65777 non-null  int64  
 1   Status                   65777 non-null  object 
 2   Location                 51672 non-null  object 
 3   City                     1383 non-null   object 
 4   Country                  65777 non-null  object 
 5   Age                      2988 non-null   object 
 6   Gender                   2473 non-null   object 
 7   Date_onset               77 non-null     object 
 8   Date_confirmation        65053 non-null  object 
 9   Symptoms                 220 non-null    object 
 10  Hospitalised (Y/N/NA)    354 non-null    object 
 11  Date_hospitalisation     35 non-null     object 
 12  Isolated (Y/N/NA)        492 non-null    object 
 13  Date_isolation           16 non-null     object 
 14  Outcome               

## Listing down of issues with the dataset that need to be cleaned

In [ ]:
# Inconsistent capitalization of "Gender" column
print('Gender:')
print(df.Gender.value_counts())

# Dates are in string form, not datetime
print(f'\nDate_onset:  {type(df.Date_onset.iloc[0])}')
print(f'Date_confirmation:  {type(df.Date_confirmation.iloc[0])}')
print(f'Date_hospitalisation:  {type(df.Date_hospitalisation.iloc[0])}')
print(f'Date_isolation:  {type(df.Date_isolation.iloc[0])}')
print(f'Date_death:  {type(df.Date_death.iloc[13204])}')
print(f'Date_entry:  {type(df.Date_entry.iloc[0])}')
print(f'Date_last_modified:  {type(df.Date_last_modified.iloc[0])}')

# Columns with semantics
print('Symptoms')
print('Contact_comment')

# Columns with inconsistent data
print('Travel_history_entry') # Inconsistent data type
print('Travel_history_start') # Inconsistent data type
print('Travel_history_location') # Contains sentences
print('Travel_history_country') # Contains sentences
print('Confirmation_method') # Contains sentences

# Unknown columns - idk what the purpose is
print('Genomics_Metadata') #i think this is important for treatment kasi and tracing where infections occured - franz
print('Source') #kinda weird nga for analysis, not super impt for us so I suggest to drop
print('Source_II') #kinda weird nga for analysis, not super impt for us so I suggest to drop

Gender:
male       2420
female       31
Male         15
male          5
female        1
Female        1
Name: Gender, dtype: int64

Date_onset:  <class 'str'>
Date_confirmation:  <class 'str'>
Date_hospitalisation:  <class 'str'>
Date_isolation:  <class 'str'>
Date_death:  <class 'str'>
Date_entry:  <class 'str'>
Date_last_modified:  <class 'str'>
Symptoms
Contact_comment
Travel_history_entry
Travel_history_start
Travel_history_location
Travel_history_country
Confirmation_method
Genomics_Metadata
Source
Source_II


In [ ]:
df['Genomics_Metadata'].value_counts() # fix capital letters to get rid of duplicates when aggregating
# retained links as they lead to relevant information regarding different genomics of the Monkeypox virus

West African Clade                                                                                                                           13
West African clade                                                                                                                            2
https://www.ncbi.nlm.nih.gov/nuccore/ON563414                                                                                                 1
Genomic sequence                                                                                                                              1
https://www.ncbi.nlm.nih.gov/nuccore/ON631963.1/                                                                                              1
https://virological.org/t/first-german-genome-sequence-of-monkeypox-virus-associated-to-multi-country-outbreak-in-may-2022/812#post_1         1
https://www.argentina.gob.ar/noticias/salud-informa-que-el-resultado-de-la-muestra-pcr-tomada-por-anlis-malbran-al-primer-caso          

## Cleaning

### Datetime


In [ ]:
# Listing all values that should be converted to datetime
# Iterate through all the columns and use the .to_datetime() of pandas to convert it to datetime type
dates = ['Date_onset', 'Date_confirmation', 'Date_hospitalisation', 'Date_isolation', 'Date_death', 'Date_entry', 'Date_last_modified']
for x in dates:
    df[x] = pd.to_datetime(df[x])


In [ ]:
df.info() #check if transformation worked

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65777 entries, 0 to 65776
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   ID                       65777 non-null  int64         
 1   Status                   65777 non-null  object        
 2   Location                 51672 non-null  object        
 3   City                     1383 non-null   object        
 4   Country                  65777 non-null  object        
 5   Age                      2988 non-null   object        
 6   Gender                   2473 non-null   object        
 7   Date_onset               77 non-null     datetime64[ns]
 8   Date_confirmation        65053 non-null  datetime64[ns]
 9   Symptoms                 220 non-null    object        
 10  Hospitalised (Y/N/NA)    354 non-null    object        
 11  Date_hospitalisation     35 non-null     datetime64[ns]
 12  Isolated (Y/N/NA)        492 non

### Fix capitals

In [ ]:
# Removing whitespace and changing the capitalisation of letters in columns that had duplicates due to different ways of typing
df['Gender'] = df['Gender'].apply(lambda x : str(x).lower().strip()) 

In [ ]:
df['Gender'].value_counts()

nan       63304
male       2440
female       33
Name: Gender, dtype: int64

In [ ]:
df['Genomics_Metadata'].apply(lambda x : str(x).lower().strip()).value_counts()

nan                                                                                                                                          65753
west african clade                                                                                                                              15
https://www.ncbi.nlm.nih.gov/nuccore/on563414                                                                                                    1
genomic sequence                                                                                                                                 1
https://www.ncbi.nlm.nih.gov/nuccore/on631963.1/                                                                                                 1
https://virological.org/t/first-german-genome-sequence-of-monkeypox-virus-associated-to-multi-country-outbreak-in-may-2022/812#post_1            1
https://www.argentina.gob.ar/noticias/salud-informa-que-el-resultado-de-la-muestra-pcr-tomada-por-anlis-malbran-al-pri

### Dropping irrelevant columns

In [ ]:
# dropping columns that aren't too relevant for analysis in this case
df.drop(['Source', 'Source_II'], axis=1, inplace=True)

In [ ]:
df

ID     Status                             Location    City  \
0          1  confirmed  Guy's and St Thomas Hospital London  London   
1          2  confirmed  Guy's and St Thomas Hospital London  London   
2          3  confirmed                               London  London   
3          4  confirmed                               London  London   
4          5  confirmed                               London  London   
...      ...        ...                                  ...     ...   
65772  65773  confirmed                             Santa Fe     NaN   
65773  65774  confirmed                     Tierra del Fuego     NaN   
65774  65775  confirmed                                  NaN     NaN   
65775  65776  confirmed                                  NaN     NaN   
65776  65777  confirmed                                  NaN     NaN   

           Country  Age Gender Date_onset Date_confirmation        Symptoms  \
0          England  NaN    nan 2022-04-29        2022-05-06            rash   
1          England  NaN    nan 2022-05-05        2022-05-12            rash   
2          England  NaN    nan 2022-04-30        2022-05-13  vesicular rash   
3          England  NaN   male        NaT        2022-05-15  vesicular rash   
4          England  NaN   male        NaT        2022-05-15  vesicular rash   
...            ...  ...    ...        ...               ...             ...   
65772    Argentina  NaN    nan        NaT        2022-09-20             NaN   
65773    Argentina  NaN    nan        NaT        2022-09-20             NaN   
65774  Puerto Rico  NaN    nan        NaT        2022-09-22             NaN   
65775  Puerto Rico  NaN    nan        NaT        2022-09-22             NaN   
65776  Puerto Rico  NaN    nan        NaT        2022-09-22             NaN   

      Hospitalised (Y/N/NA) Date_hospitalisation Isolated (Y/N/NA)  \
0                         Y           2022-05-04                 Y   
1                         Y           2022-05-06                 Y   
2                         N                  NaT                 Y   
3                         Y                  NaT                 Y   
4                         Y                  NaT                 Y   
...                     ...                  ...               ...   
65772                   NaN                  NaT               NaN   
65773                   NaN                  NaT               NaN   
65774                   NaN                  NaT               NaN   
65775                   NaN                  NaT               NaN   
65776                   NaN                  NaT               NaN   

      Date_isolation Outcome Date_death                  Contact_comment  \
0         2022-05-04     NaN        NaT                              NaN   
1         2022-05-09     NaN        NaT  Index Case of household cluster   
2                NaT     NaN        NaT                              NaN   
3                NaT     NaN        NaT              Under investigation   
4                NaT     NaN        NaT              Under investigation   
...              ...     ...        ...                              ...   
65772            NaT     NaN        NaT                              NaN   
65773            NaT     NaN        NaT                              NaN   
65774            NaT     NaN        NaT                              NaN   
65775            NaT     NaN        NaT                              NaN   
65776            NaT     NaN        NaT                              NaN   

       Contact_ID Contact_location Travel_history (Y/N/NA)  \
0             NaN              NaN                       Y   
1             3.0        Household                       N   
2             2.0        Household                       N   
3             NaN              NaN                       N   
4             NaN              NaN                       N   
...           ...              ...                     ...   
65772   

### Cleaning stray values in columns

In this part we go through the columns that had different text data as some had repeating but similar values and sentences 

We manually edit some of these values and replace it with just one datapoint so it is uniform and ready for aggregation purposes

#### Travel Location

In [ ]:
df['Travel_history_location'].value_counts()

Europe                                                                     34
Canary Islands                                                              6
international travel                                                        5
Berlin                                                                      4
London                                                                      4
Sao Paulo                                                                   3
Africa                                                                      3
Dubai                                                                       2
Nigeria                                                                     2
Spain, Britain                                                              2
USA                                                                         2
Internationally                                                             2
United Kingdom                                                  

In [ ]:
df['Travel_history_location'] = df['Travel_history_location'].replace(['Traveled out of state, within the US','out of state travel','Traveled out of state'],'out of state')
df['Travel_history_location'] = df['Travel_history_location'].replace(['Internationally', 'international travel'],'International')
df['Travel_history_location'] = df['Travel_history_location'].replace(['European countries'], 'Europe')
df['Travel_history_location'] = df['Travel_history_location'].replace(['European countries'], 'Europe')

# Replace similar values to the same value

In [ ]:
df['Travel_history_location'].value_counts() # check if it worked

Europe                                                                     35
International                                                               8
Canary Islands                                                              6
Berlin                                                                      4
London                                                                      4
out of state                                                                4
Sao Paulo                                                                   3
Africa                                                                      3
Dubai                                                                       2
United Kingdom                                                              2
Nigeria                                                                     2
Spain, Britain                                                              2
USA                                                             

#### Confirmation method

In [ ]:
df['Confirmation_method'] = df['Confirmation_method'].replace(['confirmed positive for Monkeypox via Reverse Transcriptase Polymerase Chain Reaction or RT-PCR by the DOH Research Institute for Tropical Medicine (RITM) on August 18, 2022','confirmed positive for Monkeypox via RT-PCR by the DOH RITM on August 19, 2022', 'confirmed positive for Monkeypox via Reverse Transcriptase Polymerase Chain Reaction or RT-PCR by the DOH', 'confirmed positive for Monkeypox via RT-PCR by the DOH RITM on August 19, 2022'], 'RT-PCR')

In [ ]:
df['Confirmation_method'].value_counts()

RT-PCR                             52
qPCR or RT-PCR                     32
PCR                                11
"Molecular and gene sequencing"     1
Electron microscopy                 1
Sequencing                          1
Microbiological Test                1
PCR and sequencing                  1
Name: Confirmation_method, dtype: int64

#### Travel history country

In [ ]:
df['Travel_history_country'] = df['Travel_history_country'].replace(['Spain; Portugal'], 'Spain, Portugal')

In [ ]:
df['Travel_history_country'].value_counts()

Spain                                            20
Germany                                           7
United States                                     7
England                                           7
Brazil                                            6
Portugal                                          5
USA                                               5
Nigeria                                           3
France                                            3
Mexico                                            2
Italy                                             2
UAE                                               2
Spain, Mexico                                     2
Spain, Portugal                                   2
Bangladesh                                        2
Canada                                            2
Spain, Singapore                                  1
Switzerland                                       1
United States, Canada, and Phillipeans            1
Netherlands 

#### Travel history entry

In [ ]:
df['Travel_history_entry'].value_counts()

2022-05-16          2
2022-07-01          2
2022-06-20          2
2022-07-21          2
2022-06-30          2
2022-05-04          1
2022-07-19          1
2022-07-03          1
2022-07-10          1
2022-07-13          1
2022-08-02          1
2022-07-07          1
2022-08-11          1
abroad              1
2022-08-21          1
2022-08-26          1
2022-09-10          1
2022-09-14          1
2022-06-27          1
2022-06-26          1
2022-06-28          1
2022-05-22          1
2022-05-20          1
2022-05-15          1
2022-05-27          1
2022-05-25          1
2022-05-10          1
American tourist    1
May                 1
early May           1
2022-06-12          1
2022-06-11          1
2022-06-06          1
domestic travel     1
2022-06-21          1
2022-06-09          1
2022-09-13          1
Name: Travel_history_entry, dtype: int64

In [ ]:
df['Travel_history_entry'] = df['Travel_history_entry'].replace(['early May', 'May'], '2022-05-01')
df['Travel_history_entry'] = df['Travel_history_entry'].replace(['American tourist', 'abroad', 'domestic travel'], 'NaT')

In [ ]:
df['Travel_history_entry'].value_counts()

NaT           3
2022-06-30    2
2022-05-16    2
2022-05-01    2
2022-06-20    2
2022-07-01    2
2022-07-21    2
2022-07-19    1
2022-07-03    1
2022-07-10    1
2022-07-13    1
2022-05-04    1
2022-07-07    1
2022-08-11    1
2022-08-21    1
2022-08-26    1
2022-09-10    1
2022-09-14    1
2022-08-02    1
2022-06-28    1
2022-06-27    1
2022-06-26    1
2022-06-09    1
2022-06-21    1
2022-06-06    1
2022-06-11    1
2022-06-12    1
2022-05-22    1
2022-05-10    1
2022-05-25    1
2022-05-27    1
2022-05-15    1
2022-05-20    1
2022-09-13    1
Name: Travel_history_entry, dtype: int64

In [ ]:
df['Travel_history_entry'] = pd.to_datetime(df['Travel_history_entry']) #final step same with the other values where we convert this to datetime

In [ ]:
df['Travel_history_entry'].value_counts()

2022-06-30    2
2022-05-16    2
2022-05-01    2
2022-06-20    2
2022-07-21    2
2022-07-01    2
2022-07-19    1
2022-07-03    1
2022-07-10    1
2022-07-13    1
2022-05-04    1
2022-08-02    1
2022-08-11    1
2022-08-21    1
2022-08-26    1
2022-09-10    1
2022-09-14    1
2022-07-07    1
2022-06-28    1
2022-06-27    1
2022-06-26    1
2022-06-09    1
2022-06-21    1
2022-06-06    1
2022-06-11    1
2022-06-12    1
2022-05-22    1
2022-05-10    1
2022-05-25    1
2022-05-27    1
2022-05-15    1
2022-05-20    1
2022-09-13    1
Name: Travel_history_entry, dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65777 entries, 0 to 65776
Data columns (total 28 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   ID                       65777 non-null  int64         
 1   Status                   65777 non-null  object        
 2   Location                 51672 non-null  object        
 3   City                     1383 non-null   object        
 4   Country                  65777 non-null  object        
 5   Age                      2988 non-null   object        
 6   Gender                   65777 non-null  object        
 7   Date_onset               77 non-null     datetime64[ns]
 8   Date_confirmation        65053 non-null  datetime64[ns]
 9   Symptoms                 220 non-null    object        
 10  Hospitalised (Y/N/NA)    354 non-null    object        
 11  Date_hospitalisation     35 non-null     datetime64[ns]
 12  Isolated (Y/N/NA)        492 non

#### Travel history start

In [ ]:
df['Travel_history_start'].value_counts()

late April    2
2022-04-28    1
2022-05-23    1
2022-06-09    1
2022-06-16    1
2022-06-11    1
May           1
Early july    1
2022-08-15    1
2022-08-26    1
Name: Travel_history_start, dtype: int64

In [ ]:
# Replace text data with first day of the month
df['Travel_history_start'] = df['Travel_history_start'].replace(['May'], '2022-05-01')
df['Travel_history_start'] = df['Travel_history_start'].replace(['Early july'], '2022-07-01')
df['Travel_history_start'] = df['Travel_history_start'].replace(['late April'], '2022-04-15')

In [ ]:
df['Travel_history_start'].value_counts()

2022-04-15    2
2022-04-28    1
2022-05-23    1
2022-06-09    1
2022-06-16    1
2022-06-11    1
2022-05-01    1
2022-07-01    1
2022-08-15    1
2022-08-26    1
Name: Travel_history_start, dtype: int64

In [ ]:
df['Travel_history_start'] = pd.to_datetime(df['Travel_history_start']) #final step same with the other values where we convert this to datetime

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65777 entries, 0 to 65776
Data columns (total 28 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   ID                       65777 non-null  int64         
 1   Status                   65777 non-null  object        
 2   Location                 51672 non-null  object        
 3   City                     1383 non-null   object        
 4   Country                  65777 non-null  object        
 5   Age                      2988 non-null   object        
 6   Gender                   65777 non-null  object        
 7   Date_onset               77 non-null     datetime64[ns]
 8   Date_confirmation        65053 non-null  datetime64[ns]
 9   Symptoms                 220 non-null    object        
 10  Hospitalised (Y/N/NA)    354 non-null    object        
 11  Date_hospitalisation     35 non-null     datetime64[ns]
 12  Isolated (Y/N/NA)        492 non

## Final Cleaned CSV
Export final cleaned dataset

In [ ]:
df.to_csv('Monkeypox_cleaned_data.csv')

# ISCS 30.18 Exercise 4 - Filling in Missing Data

In [ ]:
print(f'TOTAL ROW COUNT: {len(df)}\n\n')

print('MISSING ROWS COUNT:')
for i in df.columns:
    print(f'{i}: {df[i].isna().sum()}')

TOTAL ROW COUNT: 65777

ID: 0
Status: 0
Location: 14105
City: 64394
Country: 0
Age: 62789
Gender: 0
Date_onset: 65700
Date_confirmation: 724
Symptoms: 65557
Hospitalised (Y/N/NA): 65423
Date_hospitalisation: 65742
Isolated (Y/N/NA): 65285
Date_isolation: 65761
Outcome: 65752
Date_death: 65757
Contact_comment: 65686
Contact_ID: 65750
Contact_location: 65771
Travel_history (Y/N/NA): 65412
Travel_history_entry: 65738
Travel_history_start: 65766
Travel_history_location: 65663
Travel_history_country: 65678
Genomics_Metadata: 65753
Confirmation_method: 65677
Date_entry: 0
Date_last_modified: 10000


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65777 entries, 0 to 65776
Data columns (total 28 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   ID                       65777 non-null  int64         
 1   Status                   65777 non-null  object        
 2   Location                 51672 non-null  object        
 3   City                     1383 non-null   object        
 4   Country                  65777 non-null  object        
 5   Age                      2988 non-null   object        
 6   Gender                   65777 non-null  object        
 7   Date_onset               77 non-null     datetime64[ns]
 8   Date_confirmation        65053 non-null  datetime64[ns]
 9   Symptoms                 220 non-null    object        
 10  Hospitalised (Y/N/NA)    354 non-null    object        
 11  Date_hospitalisation     35 non-null     datetime64[ns]
 12  Isolated (Y/N/NA)        492 non

## Columns that have missing values

* Location
* City
* Country
* Age
* Date_onset
* Date_confirmation
* Symptoms
* Hospitalized (Y/N/NA)
* Date_isolation
* Outcome
* Date_death
* Contact_comment
* Contact_ID
* Contact_location
* Travel_history (Y/N/NA)
* Travel_history_entry
* Travel_history_start
* Travel_history_location
* Travel_history_country
* Genomics_Metadata
* Date_last_modified
* Confirmation_method

Before we proceed with handling the missing values, we set here that the expectation for processing it is for the use of a machine learning algorithm. 

The initial response to handling missing values is to drop the column entirely however, this can't be done in any of our columns except for the `Contact_comment` column because we'll be dropping highly valuable data that is important despite it being significantly less represented compared to the entire dataset. 

Imputation is another option to 

Thus, we'll group the data into relevant machine learning applications then decide how to process them from there.

This assumes that for a specific analysis, the other columns for different applications aren't relevant and thus can be dropped due to their missing values.

In this next section, we also put alongside it the  decision for handling missing values for each column

### Geospatial analysis
* Location 
* City
* Country
* Contact_location
* Travel_history (Y/N/NA)
* Travel_history_entry
* Travel_history_start
* Travel_history_location
* Travel_history_country

### Clustering
* Age
* Symptoms
* Hospitalized (Y/N/NA)
* Outcome
* Genomics_Metadata


### Natural Language Processing
* `Contact_comment` - drop the column, although this has potential to be used for NLP, at it's current state it doesn't provide too much value for ML applications of it as there is no discernable sentiment and the descriptions can be somewhat derived from other aspects of the data i.e. `status` showing that it's not confirmed indicating that it is under investigation, or index case description being derived from presence of `contact ID` in other cases

In [ ]:
df['Contact_comment'].value_counts()

Under investigation                                                                          9
Thought to be linked to cases in Canary Islands                                              6
known exposure to someone with MPX                                                           5
under investigation                                                                          5
Not related to other Mettmann district cases                                                 3
                                                                                            ..
Contact of 1266, not family                                                                  1
Contact of 1266, not family                                                                  1
Partner is negative                                                                          1
DOH identified connections between the first 5 Hawaii cases (IDs: 1078, 1562, 2007, 2008)    1
Came in contact with a person who tested positive 

In [ ]:
df.drop('Contact_comment', axis=1, inplace=True)

In [ ]:
df.columns

Index(['ID', 'Status', 'Location', 'City', 'Country', 'Age', 'Gender',
       'Date_onset', 'Date_confirmation', 'Symptoms', 'Hospitalised (Y/N/NA)',
       'Date_hospitalisation', 'Isolated (Y/N/NA)', 'Date_isolation',
       'Outcome', 'Date_death', 'Contact_ID', 'Contact_location',
       'Travel_history (Y/N/NA)', 'Travel_history_entry',
       'Travel_history_start', 'Travel_history_location',
       'Travel_history_country', 'Genomics_Metadata', 'Confirmation_method',
       'Date_entry', 'Date_last_modified'],
      dtype='object')

Processed it in the main dataframe since it doesn't necessarily affect other columns

### Timeseries Analysis
* Date_onset
* Date_confirmation
* Date_isolation
* Date_death
* Date_last_modified - since Date_entry is complete, we can assume that Date_last_modified's default value is the same as Date_entry

In [ ]:
df['Date_last_modified'].value_counts()

2022-08-16    2720
2022-08-22    2113
2022-07-26    2080
2022-08-10    1929
2022-08-08    1783
              ... 
2022-09-05       1
2022-09-04       1
2022-09-03       1
2022-07-04       1
2022-09-09       1
Name: Date_last_modified, Length: 108, dtype: int64

In [ ]:
# df['c'] = df.apply(
#     lambda row: row['a']*row['b'] if np.isnan(row['c']) else row['c'],
#     axis=1
# )

#source: https://datascience.stackexchange.com/questions/17769/how-to-fill-missing-value-based-on-other-columns-in-pandas-dataframe

df['Date_last_modified'] = df.apply(
    lambda row: row['Date_entry'] if pd.isnull(row['Date_last_modified']) else row['Date_last_modified'],
    axis = 1
)

In [ ]:
df['Date_last_modified']

0       2022-05-18
1       2022-05-18
2       2022-05-18
3       2022-05-18
4       2022-05-18
           ...    
65772   2022-09-22
65773   2022-09-22
65774   2022-09-22
65775   2022-09-22
65776   2022-09-22
Name: Date_last_modified, Length: 65777, dtype: datetime64[ns]

We can see here that there are now 65,777 rows instead of 100+ at the start. Did this in the main df since it can be considered for all ML applications.

### Network Analysis
* Contact_ID

This column connects multiple rows for Network Analysis so to do it, we must drop the other columns with null values first and then only keep rows that contain values for `Contact_ID` as well as the `ID` that is considered the `Contact_ID`

In [ ]:
df_network_analysis = df.drop(['Location', 'City', 'Country', 'Age', 'Gender', 'Date_onset',
        'Date_confirmation', 'Symptoms', 'Hospitalised (Y/N/NA)', 
        'Date_hospitalisation', 'Isolated (Y/N/NA)', 'Date_isolation',
        'Outcome', 'Date_death', 'Contact_location',
        'Travel_history (Y/N/NA)', 'Travel_history_entry',
        'Travel_history_start', 'Travel_history_location',
        'Travel_history_country', 'Genomics_Metadata', 'Confirmation_method',
        'Date_last_modified'], 
        axis = 1
        )

In [ ]:
df_network_analysis

ID     Status  Contact_ID Date_entry
0          1  confirmed         NaN 2022-05-18
1          2  confirmed         3.0 2022-05-18
2          3  confirmed         2.0 2022-05-18
3          4  confirmed         NaN 2022-05-18
4          5  confirmed         NaN 2022-05-18
...      ...        ...         ...        ...
65772  65773  confirmed         NaN 2022-09-22
65773  65774  confirmed         NaN 2022-09-22
65774  65775  confirmed         NaN 2022-09-22
65775  65776  confirmed         NaN 2022-09-22
65776  65777  confirmed         NaN 2022-09-22

[65777 rows x 4 columns]

In [98]:
df_network_analysis = df_network_analysis[pd.notnull(df_network_analysis['Contact_ID'])]

In [99]:
df_network_analysis

ID     Status  Contact_ID Date_entry
1          2  confirmed         3.0 2022-05-18
2          3  confirmed         2.0 2022-05-18
98        99  confirmed       100.0 2022-05-19
99       100  confirmed        99.0 2022-05-19
1077    1078  confirmed      1562.0 2022-06-03
1103    1104  confirmed       996.0 2022-06-03
1123    1124  confirmed      1125.0 2022-06-04
1124    1125  confirmed      1124.0 2022-06-04
1561    1562  confirmed      1790.0 2022-06-08
1707    1708  confirmed      1266.0 2022-06-10
1708    1709  confirmed      1266.0 2022-06-10
1789    1790  confirmed      2007.0 2022-06-10
1986    1987  confirmed      1266.0 2022-06-14
2006    2007  confirmed      2008.0 2022-06-14
2007    2008  confirmed      1078.0 2022-06-14
2360    2361  confirmed      2362.0 2022-06-16
2361    2362  confirmed      2363.0 2022-06-16
4515    4516  confirmed      1078.0 2022-06-24
4658    4659  confirmed      4658.0 2022-06-27
5224    5225  confirmed      4542.0 2022-06-29
5225    5226  confirmed      4542.0 2022-06-29
20262  20263  confirmed     15790.0 2022-07-27
20264  20265  suspected     15790.0 2022-07-27
20265  20266  suspected     20264.0 2022-07-27
23617  23618  confirmed     23619.0 2022-08-01
23618  23619  confirmed     23618.0 2022-08-01
27577  27578  confirmed     25755.0 2022-08-05

# Sources

* https://datascience.stackexchange.com/questions/17769/how-to-fill-missing-value-based-on-other-columns-in-pandas-dataframe


